src/app/test/utils/getIP.ts

```diff
@@ -1,6 +1,189 @@
+ * Supports both IMDS and external metadata providers.
+ */
+const networkConfig: NetworkMetadataConfig = Object.freeze({
+  publicIPv4: process.env.PUBLIC_IPV4_ADDRESS,
+  publicIPv6: process.env.PUBLIC_IPV6_ADDRESS,
+  ipv4Provider: process.env.IPV4_ADDRESS_PROVIDER,
+  ipv6Provider: process.env.IPV6_ADDRESS_PROVIDER,
+  defaultIPv4Provider: process.env.IPV4_ADDRESS_PROVIDER ?? 'http://169.254.169.254/latest/meta-data/public-ipv4',
+  defaultIPv6Provider: process.env.IPV6_ADDRESS_PROVIDER ?? 'http://169.254.169.254/latest/meta-data/ipv6',
+  imdsTokenProvider: 'http://169.254.169.254/latest/api/token',
+})
+
+/**
+ * Cached IMDS token and its expiration timestamp.
+ */
+let imdsAuthTokenCache: { value: string; expiresAt: number } | null = null
+
+/**
+ * Configures the HTTP and HTTPS client to use the specified IP family (IPv4 or IPv6).
+ *
+ * This modifies `httpClient.defaults`, meaning all subsequent requests will
+ * use the selected IP family unless changed again.
+ *
+ * @param family - The IP family (4 for IPv4, 6 for IPv6).
+ */
+const configureHttpFamily = (family: 4 | 6): void => {
+  httpClient.defaults.httpAgent = new HttpAgent({ family })
+  httpClient.defaults.httpsAgent = new HttpsAgent({ family, rejectUnauthorized: false })
+  httpClient.defaults.family = family
+}
+
+/**
+ * Sends an HTTP GET request to the specified address.
+ *
+ * @param url - The URL to request.
+ * @param headers - Optional headers for the request.
+ * @param method - Optional HTTP method
+ * @returns The trimmed response data.
+ * @throws If the response data is empty or missing.
+ */
+const fetchData = async (
+  url: string,
+  headers: Record<string, string> = {},
+  method: 'GET' | 'POST' | 'PUT' | 'PATCH' = 'GET'
+): Promise<string> => {
+  const response = await httpClient.request<string>({ url, method, headers })
+  const data = response.data?.trim()
+
+  if (!data || data.length === 0) {
+    throw new Error(`fetchData failed: ${JSON.stringify({ url, headers, method })}`)
+  }
+
+  return data
 }
+
+/**
+ * Sends an HTTP GET request using an IMDS token for authentication.
+ *
+ * If the token is expired or invalid, it will be refreshed.
+ *
+ * @param address - The URL to request.
+ * @returns The trimmed response data.
+ */
+const fetchWithIMDSToken = async (address: string): Promise<string> => {
+  if (!imdsAuthTokenCache || Date.now() >= imdsAuthTokenCache.expiresAt) {
+    await ensureValidIMDSToken()
+  }
+
+  try {
+    return fetchData(address, { 'X-aws-ec2-metadata-token': imdsAuthTokenCache!.value })
+  } catch (error) {
+    if ((error as any)?.response?.status === 401) {
+      console.warn(`IMDS token rejected, refreshing and retrying...`)
+      await ensureValidIMDSToken()
+      return fetchData(address, { 'X-aws-ec2-metadata-token': imdsAuthTokenCache!.value })
+    }
+    throw error
+  }
+}
+
+/**
+ * Retrieves the public IP address (IPv4 or IPv6) using the best available source.
```

Comment:
We should put comments when necessary, but we should not comment every function we create because it creates maintenance work for us. When we change the code, we need to also change the comment. Comments aren't tested, they get easily outdated. 

I believe you can keep the comments for env variables, but remove all others.

Reply:
Updated, and there are still some remaining comments, I believe they are necessary